File: haushaltsbuch/src/main/java/org/github/haschi/haushaltsbuch/api/_Inventar.java
package org.github.haschi.haushaltsbuch.api;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.github.haschi.haushaltsbuch.infrastruktur.modellierung.de.Information;
import org.immutables.value.Value;
import org.javamoney.moneta.function.MonetaryFunctions;

import java.util.List;

@Information
@JsonSerialize(as = Inventar.class)
@JsonDeserialize(as = Inventar.class)
public abstract class _Inventar
{
    public static Inventar leer()
    {
        return Inventar.builder()
                .build();
    }

    public abstract Vermögenswerte anlagevermoegen();

    public abstract Vermögenswerte umlaufvermoegen();

    public abstract Schulden schulden();

    @Value.Derived
    @JsonIgnore
    public Reinvermögen reinvermoegen()
    {
        final Währungsbetrag anlagevermögen = anlagevermoegen().summe();
        final Währungsbetrag umlaufvermögen = umlaufvermoegen().summe();
        final Währungsbetrag schulden = schulden().summe();

        return Reinvermögen.builder()
                .summeDesVermögens(Währungsbetrag.of(anlagevermögen.wert().add(umlaufvermögen.wert())))
                .summeDerSchulden(schulden)
                .build();
    }
}

File: haushaltsbuch/src/main/kotlin/org/github/haschi/haushaltsbuch/core/Inventar.kt
package org.github.haschi.haushaltsbuch.core

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.github.haschi.haushaltsbuch.api.Reinvermögen
import org.github.haschi.haushaltsbuch.api.Schulden
import org.github.haschi.haushaltsbuch.api.Vermoegenswerte
import org.github.haschi.haushaltsbuch.api.Währungsbetrag

@JsonSerialize(`as` = Inventar::class)
@JsonDeserialize(`as` = Inventar::class)
@JsonIgnoreProperties(value= "reinvermoegen", allowGetters=true)
data class Inventar(
        val anlagevermoegen: Vermoegenswerte,
        val umlaufvermoegen: Vermoegenswerte,
        val schulden: Schulden)
{

    val reinvermoegen: Reinvermögen
        get() = Reinvermögen(
                Währungsbetrag(
                        anlagevermoegen.summe.wert.add(
                                umlaufvermoegen.summe.wert)),
                schulden.summe)


    companion object
    {
        val leer: Inventar
            get() = Inventar(
                    Vermoegenswerte.keine,
                    Vermoegenswerte.keine,
                    Schulden.keine)
    }
}
